raft.go

```diff
@@ -1907,6 +2044,82 @@ func (r *Raft) electSelf() <-chan *voteResult {
 	return respCh
 }
 
+// preElectSelf is used to send a RequestPreVote RPC to all peers, and vote for
+// ourself. This will not increment the current term. The
+// response channel returned is used to wait for all the responses (including a
+// vote for ourself).
+// This must only be called from the main thread.
+func (r *Raft) preElectSelf() <-chan *preVoteResult {
+	// Create a response channel
+	respCh := make(chan *preVoteResult, len(r.configurations.latest.Servers))
+
+	// Increment the term
+	newTerm := r.getCurrentTerm() + 1
+
+	// Construct the request
+	lastIdx, lastTerm := r.getLastEntry()
+	req := &RequestPreVoteRequest{
+		RPCHeader:    r.getRPCHeader(),
+		Term:         newTerm,
+		LastLogIndex: lastIdx,
+		LastLogTerm:  lastTerm,
+	}
+
+	// Construct a function to ask for a vote
+	askPeer := func(peer Server) {
+		r.goFunc(func() {
+			defer metrics.MeasureSince([]string{"raft", "candidate", "electSelf"}, time.Now())
+			resp := &preVoteResult{voterID: peer.ID}
+			if prevoteTrans, ok := r.trans.(WithPreVote); ok {
```

Comment:
nit: would it be cleaner to do this type assertion right at the top of `preElectSelf` and bail out immediately if the transport doesn't support it (which _shouldn't_ be possible due to the check in NewRaft but defensive against future change...).

Main rationale is that it allows un-indenting and simplifying the body of this callback which makes it easier to follow the important parts!

Reply:
I changed that and added a defensive panic. Please check if this is what you have in mind.